plugin/core/views.py

```diff
@@ -172,3 +176,39 @@ def text_document_range_formatting(view: sublime.View, region: sublime.Region) -
         "options": formatting_options(view.settings()),
         "range": region_to_range(view, region).to_lsp()
     })
+
+
+def minihtml(content: Union[str, dict], view: sublime.View) -> str:
+    """ Content can be a string or MarkupContent """
+    if isinstance(content, str):
+        return text2html(content)
+    elif isinstance(content, dict):
+        value = content.get("value") or ""
+        kind = content.get("kind")
+        if kind == "markdown":
+            return mdpopups.md2html(view, value)
+        else:
+            # must be plaintext
+            return text2html(value)
+    else:
+        return ''
+
+
+def text2html(content: str) -> str:
+    content = "\n".join(textwrap.wrap(content, width=54,
+                                      replace_whitespace=False, break_long_words=False, break_on_hyphens=False))
+    content = html.escape(content).replace('\n', '<br>').replace('\t', '&nbsp;&nbsp;&nbsp;&nbsp;')
+
+    def replace_npbs(match: Any) -> str:
+        spaces = match.group(0)
+        return "&nbsp;" * len(spaces)
+
+    # if there are 2 or more spaces, replace them with &nbsp;
+    content = re.sub(r"( {2,})", replace_npbs, content)
+
+    def replace_url_with_link(match: Any) -> str:
+        url = match.group(0)
+        return "<a href='{}'>{}</a>".format(url, url)
+
+    FIND_URL = re.compile(r'(https?://(?:[\w\d:#@%/;$()~_?\+\-=\\\.&](?:#!)?)*)', flags=re.IGNORECASE)
```

Comment:
If valid MD is being passed, shouldn't md2html cover this?

Reply:
There are two content formats. One is markdown, the other is plaintext. When the server returns markdown, it is indeed handled by `mdpopups.md2html`. When the server returns plaintext, we have to put a little more thought into faithfully preserving the original plaintext when giving it to ST's minihtml.